### PR TITLE
Auto-media embedding

### LINKED ISSUE #6: Implement the auto embed (sub)feature
Just [like in v4](https://sdk.ckeditor.com/samples/mediaembed.html) we probably want to implement the Auto Embed. Long story short, it's all about upcasting the pasted urls from the clipboard straight to embedded media. Check out the v4 demo to learn more.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced auto-media embedding feature. Closes #6.

### CODE REVIEW

File: tests/mediaembed.js [hunk 1]
```diff
@@ -44,11 +49,183 @@ describe( 'MediaEmbed', () => {
...
+
 	it( 'should load MediaEmbedUI plugin', () => {
 		expect( editor.plugins.get( MediaEmbedUI ) ).to.instanceOf( MediaEmbedUI );
 	} );
 
 	it( 'has proper name', () => {
 		expect( MediaEmbed.pluginName ).to.equal( 'MediaEmbed' );
 	} );
+
+	describe.only( 'auto-media embed', () => {
```
Comment (Reinmar) [inline tests/mediaembed.js]: `only()`

File: src/mediaembed.js [hunk 2]
```diff
@@ -37,6 +44,89 @@ export default class MediaEmbed extends Plugin {
...
+	init() {
+		this._attachAutoEmbedingEvents();
+	}
+
+	/**
+	 * Attach events required for "auto-embeding" feature.
+	 *
+	 * @private
+	 */
+	_attachAutoEmbedingEvents() {
```
Comment (oleq) [inline src/mediaembed.js]: This is definitely a logic of `MediaEmbedEditing`, not `MediaEmbed`.

This could even be a separate plugin like `AutoMediaEmbed` loaded by default so even without config, users are able to get rid of it by creating a custom build.

File: src/mediaembed.js [hunk 3]
```diff
@@ -37,6 +44,89 @@ export default class MediaEmbed extends Plugin {
...
+
+			// If the url is valid from MediaEmbed plugin point of view, let's use it.
+			if ( !mediaRegistry.hasMedia( url ) ) {
+				return detach();
+			}
+
+			// `leftLivePosition` won't be available in `setTimeout` function so let's clone it.
+			const positionToInsert = Position.createFromPosition( leftLivePosition );
+
+			global.window.setTimeout( () => {
```
Comment (oleq) [inline src/mediaembed.js]: 1. Paste.
2. Type.
3. Auto-embed after 500ms.
4. But the error is thrown. Probably because some ranges from 1. are no longer valid because of 2.

kapture 2018-09-11 at 12 40 42

File: src/mediaembed.js [hunk 4]
```diff
@@ -37,6 +44,89 @@ export default class MediaEmbed extends Plugin {
...
+				editor.model.change( writer => {
+					// const mediaElement = writer.createElement( 'media', { url } );
```
Comment (oleq) [inline src/mediaembed.js]: Do we need that code?

File: src/mediaembed.js [hunk 5]
```diff
@@ -37,6 +44,89 @@ export default class MediaEmbed extends Plugin {
...
+
+					writer.remove( urlRange );
+					// writer.insert( mediaElement, positionToInsert );
```
Comment (oleq) [inline src/mediaembed.js]: Do we need that code?

File: src/mediaembed.js [hunk 6]
```diff
@@ -37,6 +44,89 @@ export default class MediaEmbed extends Plugin {
...
+			for ( const node of walker ) {
+				if ( node.type === 'elementStart' ) {
+					return detach();
+				}
+
+				url += node.item.data;
+			}
+
+			// If the url does not match to universal url regexp, let's skip that.
+			if ( !url.match( URL_REGEXP ) ) {
```
Comment (oleq) [inline src/mediaembed.js]: Paste the following "fusion" URL into the content: 
```
https://www.youtube.com/watch?v=H08tGjXNHO4https://open.spotify.com/album/2IXlgvecaDqOeF3viUZnPI?si=ogVw7KlcQAGZKK4Jz9QzvA
```

The following broken widget is created
image

Something's wrong with discovering URLs in the pasted content.

File: src/mediaembed.js (hunk 3 above)
Comment (oleq) [reply src/mediaembed.js]: Similarly

1. Paste `https://www.youtube.com/watch?v=H08tGjXNHO4`,
2. Undo,
3. Auto–embed after 500ms.

Also throws the `Uncaught CKEditorError: move-operation-nodes-do-not-exist: The nodes which should be moved do not exist.` error.

File: src/mediaembed.js (hunk 6 above)
Comment (oleq) [reply src/mediaembed.js]: Also pasting `\s*https://www.youtube.com/watch?v=H08tGjXNHO4\s*` should result in auto-embed IMO. We mustn't assume users are white-space perfect when copy&pasting stuff from other places (e.g. other documents, some lists). That would be annoying from the UX point of view.

File: src/mediaembed.js (hunk 2 above)
Comment (pomek) [reply src/mediaembed.js]: Yes, it could be moved as a separate plugin.

File: src/mediaembed.js (hunk 3 above)
Comment (pomek) [reply src/mediaembed.js]: That's the negative side of the timeout. I don't have an idea how we could solve it. Do you have any?

File: src/mediaembed.js (hunk 4 above)
Comment (pomek) [reply src/mediaembed.js]: No, it's my mistake.

File: src/mediaembed.js (hunk 6 above)
Comment (pomek) [reply src/mediaembed.js]: When I paste `https://www.youtube.com/watch?v=H08tGjXNHO4https://open.spotify.com/album/2IXlgvecaDqOeF3viUZnPI?si=ogVw7KlcQAGZKK4Jz9QzvA` nothing happens.

**Comment by pomek** [comment]
- I changed the delay from 500ms to 100ms. Sometimes I am able to reproduce an error by pressing `cmd+v` and `cmd+z` but I must do it really fast. It sounds like an edge case.
- `AutoMediaEmbed` plugin is now live. The entire logic that I wrote in `MediaEmbed` was moved to the new plugin.
- URL that contains spaces before or after (`  URL `) will be handled as well.
- `https://www.youtube.com/watch?v=H08tGjXNHO4https://open.spotify.com/album/2IXlgvecaDqOeF3viUZnPI?si=ogVw7KlcQAGZKK4Jz9QzvA` does not trigger the logic of auto-embed. Could you test it once again?

**Comment by oleq** [comment]
I'm on it, thanks.

File: src/automediaembed.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+ * The auto-media embed plugin.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class AutoMediaEmbed extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ Clipboard ];
```
Comment (oleq) [inline src/automediaembed.js]: I'm not sure we need that. Importing that is the job of the `Essentials` plugin, which is the key part of each editor.

Comment (oleq) [reply src/automediaembed.js]: OTOH, it doesn't hurt so I guess it can stay :P

File: src/automediaembed.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+			}
+
+			const urlRange = new Range( leftLivePosition, rightLivePosition );
+			const walker = new TreeWalker( { boundaries: urlRange, ignoreElementEnd: true } );
+
+			let url = '';
+
+			for ( const node of walker ) {
+				if ( node.type === 'elementStart' ) {
+					return detach();
```
Comment (oleq) [inline src/automediaembed.js]: This line is not covered by tests.

**Comment by oleq** [comment]
> I changed the delay from 500ms to 100ms. Sometimes I am able to reproduce an error by pressing `cmd+v` and `cmd+z` but I must do it really fast. It sounds like an edge case.

I'm a little bit worried about that. It may be hard to reproduce when there's just one user but when multiple users are typing around, it could throw errors very often. We need to check if it goes well with collaboration.

**Comment by oleq** [comment]
If the feature fails in obvious collaboration cases, we have to solve https://github.com/ckeditor/ckeditor5-media-embed/pull/27#discussion_r216623498 anyway.

File: src/automediaembed.js (hunk 7 above)
Comment (pomek) [reply src/automediaembed.js]: You're right when you said:

> I'm not sure we need that. Importing that is the job of the Essentials plugin, which is the key part of each editor.

but we cannot assume that every editor will use the `Essentials` plugin. I'm defining it as a dependency because it's required for `AutoMediaEmbed` to work from its point of view.

**Comment by pomek** [comment]
More tests have been added. Now they check some cases that can occur during collaboration.
